### ISSUE TITLE
GNU command line interface

### ISSUE THREAD

**Issue by pditommaso** [nextflow-io/nextflow#556]
Nextflow currently uses single `-` prefix for command level options and `--` prefix for the definition of user workflow parameters. 

Users frequently do not realise the difference and use the `--` prefix for nextflow options, which instead are interpreted workflow parameters and therefor ignored, causing confusion. 

For example people tend to use `--resume` instead of `-resume` causing the pipeline to re-start from scratch.  
  
The goal of this issue is to re-organise the nextflow command line interface using the GNU convention i.e. using a single `-` for short options (one char) and `--` for long options (2 or more chars) since this is the most common convention. 

To distinguish nextflow options from the user workflow parameters, the `run` command line needs to be interpreted in a positional manner ie. all options up from the `run` command to the project/script to be executed are interpreted as run options. Following the project/script name are interpreted as the user workflow parameters. For example: 

```
nextflow run --revision v1.1 --resume <project name> --foo x --bar y 
```

Whenever possible the new CLI should be backward compatible with the current CLI, automatically replacing any single `-` prefix to a `--` prefix and showing a warning message to the user.  

A library that's a good candidate on which base the new CLI is [Picocli](https://github.com/remkop/picocli).

**Comment by shaze** [comment]
Dear Paolo

Not a volunteer, but a suggestion of a coding style approach to dealing with this problem until this is done (which I think would be a good idea).

It's fairly easy to integrate into a pipeline -- just have a list of allowed parameters which get checked. Even for a complex pipeline it shouldn't take more than 5 minutes to do. A similar approach can be used for providing command line help.

```
allowed_params = ["start", "cutoff", "theta","foo"]

params.each { entry ->
  if (! allowed_params.contains(entry.key)) {
      println("The parameter <${entry}.key> is not known");
      System.exit(2);
  }
}    


if (params.help) {
    params.each {
    entry ->
      print "Parameter: <$entry.key>    \t Default: $entry.value"
      help = params_help.get(entry.key)
      if (help) {
          print "\n    $help"
          println ""
      }
  }
  System.exit(0)
}
```

**Comment by msmootsgi** [comment]
FWIW, I only ever use nextflow's native command line arguments on the actual command line.  All other parameters get specified with `-params-file` and a YAML file.  This has two advantages: it's easy to track what you did and it's easy to standardize the execution of arbitrary pipelines in our cloud infrastructure.  The obvious disadvantage is that it takes effort to create the YAML file.

The only time I use command line params is when I'm working with dummy code and testing ideas.

**Comment by pditommaso** [comment]
Thanks for you feedback, however it's quite handy to have the ability to pass parameters as CLI options. It should be preserved.

**Comment by pditommaso** [comment]
See also #129

**Comment by edgano** [comment]
#129 can be solved using positional parameters
_2.4. Mixing Options and Positional Parameters_ See [doc](http://picocli.info/)

**Comment by pditommaso** [comment]
The new command line is mostly implemented, however a decision needs to be taken how handle implicit option and user parameters. There are mainly two choices:

1) Use a strict positional command line option, think like `docker run [options] <image> [params]`. In the case of nextflow would be `nextflow run [option] <pipeline> [pipeline parameters]` eg: 

    ```
    nextflow run --revision foo --work-dir /usr/workspace rnaseq-nf --reads /some/path  
   ```

    The problem of this approach is that does not allow command line option implicit values, for example it won't be possibile to just say `--resume`, but it would be mandatory to provide a value for the resume option (because otherwise it could consume the pipeline name as a resume value).


2. another option is to separate NF command line option from pipeline parameters using a `--` separator. This would allow the use of a *not* positional command line (like it's now) and therefore the ability of using implicit option values. For example: 

    ```
    nextflow run rnaseq-nf --revision foo --resume -- --reads /somedata 
    ```

Even this it's a bit more verbose I'm thinking that maybe a better solution, also because impose a clearly separation from NF options and script option, helping to avoid problems like #624. 

@emi80 @msmootsgi @ewels What do you think?

**Comment by msmootsgi** [comment]
I don't mean to further confuse matters, but have you considered user parameters similar to how `java` handles extra parameters like `-Xmx10G`?  So nextflow would define all it's normal command line params and then any user defined params would be preceded with `-P`, so for `params.name` within your script you'd have `-Pname=whatever` on the command line?  That should make it very clear which params are which.

**Comment by edgano** [comment]
I think second option will be better. More readable and more powerful in the future

**Comment by ewels** [comment]
Can you not just have a list of protected nextflow vocabulary? eg. Exit with an error if anyone tries to use `params.resume`? Then try to parse the command line options with these first and anything that’s left must be a params option. Then you can be super flexible with how they’re specified..

**Comment by msmootsgi** [comment]
@ewels I like the idea, but it limits adding options/parameters to core nextflow in the future since any new nextflow options might stomp on something I added to my pipeline.

**Comment by ewels** [comment]
Yup, true. But how many new command line options have been added recently? If nextflow throws an error when `params.newcommand` is mentioned anywhere then it’ll be easy to fix and hopefully not more than an inconvenience.

I just worry that this issue is to reduce confusion by users, but the two proposals will still be confusing but just in different ways.

**Comment by shaze** [comment]
I think  I would lean towards a variant of the first option. In a way it's  there already. With the cloud command, the "-c" can occur before or after the "cloud" with different meanings. 

`nextflow -c scott.aws -c run10.config cloud create scottcluster -c 5`

**Comment by rsuchecki** [comment]
I normally strongly advocate the common `-x` `--long-x-option` approach and cringe a  bit when encountering `-long`. However in the case of NF the well-defined use of `-long` vs  `--long` clearly delineates NF opts and pipeline params. It also allows the user not to think about correctly positioning opts/params. Short opts are great for bashing away in an interactive session, trying things out - this is not what NF is about, right? 

As for the example with erroneous use of `--resume` vs `-resume`, wouldn't that be better addressed by stopping the execution if any undefined pipeline opt is encountered? 

Finally, the two approaches proposed by @pditommaso are reasonable, with the second one making it easier for the user to focus on either NF or pipeline params as needed.

**Comment by ewels** [comment]
> As for the example with erroneous use of `--resume` vs `-resume`, wouldn't that be better addressed by stopping the execution if any undefined pipeline opt is encountered?

Agreed - in fact, this could probably resolve many of the problems and questions that we see. Especially if that error message checked for similar core parameters and gave a hint.

**Comment by edgano** [comment]
I think the first idea was to throw a WARN. Telling about the new "policy" of **--long-option**
But IDK how is in this days.

**Comment by pditommaso** [comment]
> As for the example with erroneous use of --resume vs -resume, wouldn't that be better addressed by stopping the execution if any undefined pipeline opt is encountered?

This should double, I will give it a try.

**Comment by winni2k** [comment]
FWIW, I think the second option in https://github.com/nextflow-io/nextflow/issues/556#issuecomment-371577103 (using `--` to separate nextflow from user params arguments) is simple and avoids lots of headaches.  Also, this convention will be known to many programmers that also use the popular tool `git`, as that tool follows that convention as well.

**Comment by winni2k** [comment]
Github says I just unassigned @edgano. I'm not sure how I did that...

**Comment by ewels** [comment]
> this convention will be known to many programmers

FWIW I don't think I was aware of this convention before this thread. Also, many of the people who launch our pipelines aren't programmers; they're biologists who have recently learned to use the command line. I'm a little anxious that such a convention could cause more headaches rather than fewer. I still think that having very simplistic flag handling with a protected vocab is the most user-friendly (though I agree, perhaps not the most developer-friendly).

**Comment by winni2k** [comment]
*sadness consumes me* 😞

**Comment by ewels** [comment]
I'm not sure that there was ever a consensus? I mostly see several different variants suggested with proponents of each. Personally I think that any change in the cli handling now would cause massive confusion and problems for the (now-large) user base, and this far outweighs any potential gains for me.

FWIW, most of the things that I suggested above have now been implemented within the @nf-core pipeline template: reserving core Nextflow options to avoid confusion with parameters, and validation of parameter inputs including warning about unexpected parameters.
